### PR TITLE
GH-94382: port multiprocessing static types to heap types

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-94382

### CODE REVIEW

**Comment by kumaraditya303** [comment]
Refleaks bots are failing because of https://github.com/python/cpython/pull/94330

**Comment by erlend-aasland** [comment]
@kumaraditya303, thanks for the PR. One of the most important aims of PEP-687 is to _communicate_ the changes before they happen. See [Part 1: Preparation](https://peps.python.org/pep-0687/#part-1-preparation). I'd prefer if you please re-read the PEP and follow its guidelines as specified. You'll notice that opening a PR comes in the second half of the process.

**Comment by kumaraditya303** [comment]
> One of the most important aims of [PEP-687](https://peps.python.org/687) is to _communicate_ the changes before they happen. See [Part 1: Preparation](https://peps.python.org/pep-0687/#part-1-preparation). I'd prefer if you please re-read the PEP and follow its guidelines as specified. You'll notice that opening a PR comes in the second half of the process.

I'll create an issue to discuss it but I don't expect much as the module has no global state and the implemented types are not performance critical. I had worked on this months ago just was waiting for PEP to get accepted :)

**Comment by erlend-aasland** [comment]
> [...] as the module has no global state and the implemented types are not performance critical [...]

If the types do not access global module state they should remain static; this is explicitly noted in the PEP.

**Comment by erlend-aasland** [comment]
> I'll create an issue to discuss it but I don't expect much [...]

Well, the important thing is that it gets done. Lack of communication is the single one issue that has contributed to heating the discussions regarding these changes. Even if we expect few people to participate in such a topic, it is of importance to raise awareness (and then wait at least for a week or two before continuing).

File: Modules/_multiprocessing/multiprocessing.c
Comment (encukou) [inline Modules/_multiprocessing/multiprocessing.c]: The module state is unnecessary here.
With this change:
```suggestion
    PyObject *semlock_type = (PyTypeObject *)PyType_FromModuleAndSpec(
                module, &_PyMp_SemLockType_spec, NULL);
```
and adding some `Py_DECREF(semlock_type)`, you can change `state->semlock_type` to just `semlock_type` in this function, and remove `_multiprocessingstate`, `get_module_state`, `multiprocessing_traverse`, `multiprocessing_clear`, `multiprocessing_dealloc`.

Comment (kumaraditya303) [reply Modules/_multiprocessing/multiprocessing.c]: removed

**Comment by kumaraditya303** [comment]
Removing do-not-merge as no alternative PR has been proposed so far.

File: Modules/_multiprocessing/multiprocessing.c
Comment (erlend-aasland) [inline Modules/_multiprocessing/multiprocessing.c]:
```suggestion
    int rc = PyModule_AddType(module, semlock_type);
    Py_DECREF(semlock_type);
    if (rc < 0) {
        return -1;
    }
```

File: Modules/_multiprocessing/multiprocessing.c
Comment (erlend-aasland) [inline Modules/_multiprocessing/multiprocessing.c]: Since you are touching practically the whole function, please follow PEP-7.

```suggestion
    if ((int)(SEM_VALUE_MAX) < (0)) {
        py_sem_value_max = PyLong_FromLong(INT_MAX);
    }
    else {
        py_sem_value_max = PyLong_FromLong(SEM_VALUE_MAX);
    }
```

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

File: Modules/_multiprocessing/multiprocessing.c
Comment (arhadthedev) [reply Modules/_multiprocessing/multiprocessing.c]: @erlend-aasland Can we move out repeated `py_sem_value_max = PyLong_FromLong(...)` altogether?

```c
    long sem_value_max = (int)(SEM_VALUE_MAX) < 0 ? INT_MAX : SEM_VALUE_MAX;
    py_sem_value_max = PyLong_FromLong(sem_value_max);
```

Comment (erlend-aasland) [reply Modules/_multiprocessing/multiprocessing.c]: I'd prefer not to. We tend to not use the implicit `cond ? expr : expr` in the CPython code base; the explicit `if ... else ...` is preferred, mostly because of readability. (That does not mean there is no such condensed oneliners; you'll find them, but they are rare compared to the expanded version.)

**Review by erlend-aasland — APPROVED** [review]
LGTM! Let's wait for Petr's final decision.

**Comment by erlend-aasland** [comment]
> Removing do-not-merge as no alternative PR has been proposed so far.

Well, only three hours passed. That's a little narrow time-frame, don't you agree? :)

File: Modules/_multiprocessing/semaphore.c
Comment (erlend-aasland) [inline Modules/_multiprocessing/semaphore.c]: Hm, this is file-local, so it can be static (unless I missed something).

**Comment by erlend-aasland** [comment]
As of Petr's last comment, please re-label this with do-not-merge until there is an agreement.

File: Modules/_multiprocessing/semaphore.c
Comment (vstinner) [reply Modules/_multiprocessing/semaphore.c]: Please make it static.

File: Modules/_multiprocessing/multiprocessing.c
Comment (vstinner) [inline Modules/_multiprocessing/multiprocessing.c:209]: Please add an empty line for readability here.

**Review by vstinner — COMMENTED** [review]
Please add a NEWS entry explaining that the purpose of this PR is to fix memory leaks when the ``_multiprocessing`` extension module is loaded/unloaded multiple times: https://github.com/python/cpython/issues/94382#issuecomment-1173131209

**Comment by kumaraditya303** [comment]
> Please add a NEWS entry explaining that the purpose of this PR is to fix memory leaks when the _multiprocessing extension module is loaded/unloaded multiple times: https://github.com/python/cpython/issues/94382#issuecomment-1173131209

This means I can remove do-not-merge, right ?

File: Modules/_multiprocessing/multiprocessing.c
Comment (kumaraditya303) [reply Modules/_multiprocessing/multiprocessing.c:209]: done

**Comment by vstinner** [comment]
> This means I can remove do-not-merge, right ?

That's unrelated. @erlend-aasland wrote:

> As of Petr's last comment, please re-label this with do-not-merge until there is an agreement.

Since @erlend-aasland and @encukou wrote PEP 687, I understand that you need to agree on that specific PR.

**Comment by vstinner** [comment]
Sorry, I'm no longer convinced that this PR fix an actual memory leak. The NEWS entry no looks wrong to me.

Test from https://github.com/python/cpython/issues/94382#issuecomment-1172881601:

```
from test import support
import sys
for i in range(1, 10):
    support.run_in_subinterp("import multiprocessing")
    print("refs:", sys.gettotalrefcount(), "blocks: ", sys.getallocatedblocks())
```

Without this PR, Python **leaks memory**: refs +469, blocks +160.

```
refs: 162138 blocks:  47456
refs: 162201 blocks:  47476
refs: 162259 blocks:  47496
refs: 162317 blocks:  47516
refs: 162375 blocks:  47536
refs: 162433 blocks:  47556
refs: 162491 blocks:  47576
refs: 162549 blocks:  47596
refs: 162607 blocks:  47616
```

With this PR (rebased to main), Python **still leaks memory**: refs +469, blocks +160 (no change).

```
refs: 162138 blocks:  47446
refs: 162201 blocks:  47466
refs: 162259 blocks:  47486
refs: 162317 blocks:  47506
refs: 162375 blocks:  47526
refs: 162433 blocks:  47546
refs: 162491 blocks:  47566
refs: 162549 blocks:  47586
refs: 162607 blocks:  47606
```

Tell me if I did something wrong.

---

If I modify the test to import the ``_multiprocessing`` extension instead, I see no leak on the Python main branch (without this PR):

```
refs: 162674 blocks:  49018

refs: 162679 blocks:  49018
refs: 162679 blocks:  49018
refs: 162679 blocks:  49018
refs: 162679 blocks:  49018
refs: 162679 blocks:  49018
refs: 162679 blocks:  49018
refs: 162679 blocks:  49018
refs: 162679 blocks:  49018
```

Note: I added an empty line for readability.

**Comment by kumaraditya303** [comment]
I had only verified single leak https://github.com/python/cpython/issues/94382#issue-1287280621 which is fixed by this PR. If the multiple initialization is not fixed, there must be something in multiprocessing module which is causing it. Anyways, I'll wait for decision on the discourse now before putting anymore effort into this.

**Comment by erlend-aasland** [comment]
> This means I can remove do-not-merge, right ?

Please wait until there is an agreement for this change on Discourse.

File: Modules/_multiprocessing/semaphore.c
Comment (erlend-aasland) [reply Modules/_multiprocessing/semaphore.c]: This change is done; marking this as resolved.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
FYI, I'll merge this in a day or two.

**Comment by erlend-aasland** [comment]
I'm hesitant to backport this change. Let me know if you disagree, Victor / Petr.

File: Modules/_multiprocessing/semaphore.c
Comment (erlend-aasland) [inline Modules/_multiprocessing/semaphore.c:761]: The `Py_tp_getattro`, `Py_tp_setattro`, `Py_tp_alloc`, and `Py_tp_free` overrides are superfluous, right?

Comment (kumaraditya303) [reply Modules/_multiprocessing/semaphore.c:761]: Yes explicit is better than implicit :) It is the same as in functools extension module.

Comment (erlend-aasland) [reply Modules/_multiprocessing/semaphore.c:761]: I'm not sure explicit is better than implicit here ;) I'm -1 to adding them, but I won't make a big fuzz about it.

**Comment by erlend-aasland** [comment]
Thanks, Kumar!

**Pull request merged**
